fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1676,24 +1678,59 @@ private static void checkStoreHeaderInternal(@Nonnull RecordMetaDataProto.DataSt
         }
     }
 
+    private void beginRecordStoreStateRead() {
+        // When the record store state is being updated multiple times, this function (and its implicit retry loop at
+        // the atomic reference level) will retry the update on the new record store state, so the operation always
+        // does what's expected (i.e., update the "in flight reads" value while leaving the record store state otherwise
+        // in tact).
+        recordStoreStateRef.updateAndGet(state -> {
+            state.beginRead();
+            return state;
+        });
+    }
+
+    private void endRecordStoreStateRead() {
+        recordStoreStateRef.updateAndGet(state -> {
+            state.endRead();
+            return state;
+        });
+    }
+
+    private void beginRecordStoreStateWrite() {
+        recordStoreStateRef.updateAndGet(state -> {
+            state.beginWrite();
+            return state;
+        });
+    }
+
+    private void endRecordStoreStateWrite() {
+        recordStoreStateRef.updateAndGet(state -> {
+            state.endWrite();
+            return state;
+        });
+    }
+
     @Nonnull
     private CompletableFuture<RecordMetaDataProto.DataStoreInfo> loadStoreHeaderAsync(@Nonnull StoreExistenceCheck existenceCheck, @Nonnull IsolationLevel isolationLevel) {
         return readStoreFirstKey(context, getSubspace(), isolationLevel).thenApply(keyValue -> checkAndParseStoreHeader(keyValue, existenceCheck));
     }
 
     private void saveStoreHeader(@Nonnull RecordMetaDataProto.DataStoreInfo storeHeader) {
-        if (recordStoreState == null) {
+        if (recordStoreStateRef == null) {
             throw new RecordCoreException("cannot update store header with a null record store state");
         }
-        recordStoreState.beginWrite();
+        beginRecordStoreStateWrite();
         try {
             context.setDirtyStoreState(true);
-            synchronized (recordStoreState) {
-                recordStoreState.setStoreHeader(storeHeader);
+            synchronized (this) {
```

Comment:
I think I would rather this be on `recordStoreStateRef` than `this`, as I don't think we need to synchronize all operations locked on `this` here, just operations modifying the record store header. I could be wrong, I suppose.

Reply:
Hm, alright. I *think* that it would actually be fine to synchronize on the reference here (we're using the object essentially as a lock so we can update the state and the serialized state in the transaction at once rather than to prevent multiple people from updating the atomically referenced thing at once, which is I think the mistake the spotbugs warning is about). It probably doesn't actually matter given the rarity of people updating the record store state.

Reply:
> I think that it would actually be fine to synchronize on the reference here 
> ...
> It probably doesn't actually matter given the rarity of people updating the record store state.

Yeah, I agree with both. Since it's rare, lets keep it as is. But we can ignore the spotbugs too.